src/EFCore.Relational/Update/Internal/CompositeRowValueFactory.cs

```diff
@@ -19,19 +19,26 @@ public abstract class CompositeRowValueFactory
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
-    public CompositeRowValueFactory(IReadOnlyList<IColumn> columns)
+    protected CompositeRowValueFactory(IReadOnlyList<IColumn> columns)
     {
         Columns = columns;
-        EqualityComparer = CreateEqualityComparer(columns);
     }
+    
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    protected virtual List<ValueConverter?>? ValueConverters { get; set; }
 
     /// <summary>
     ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
     ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
-    public virtual IEqualityComparer<object?[]> EqualityComparer { get; }
+    public virtual IEqualityComparer<object?[]> EqualityComparer { get; set; } = null!;
```

Comment:
```suggestion
    public virtual IEqualityComparer<object?[]> EqualityComparer { get; protected set; } = null!;
```

Good candidate to use `required`